<a href="https://colab.research.google.com/github/Nikocioni/progetto_programmazione_nik_ross_salvo/blob/main/progetto_niko_ross_salvo_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
test= pd.read_csv("/content/sample_data/1.test_convertito_numerici.csv", sep=",")
test_non = pd.read_csv("/content/sample_data/1.test_non_convertito.csv", sep=",")
train_non = pd.read_csv("/content/sample_data/1.train_non_convertito.csv", sep=",")
train = pd.read_csv("/content/sample_data/1.train_convertito_numerici.csv", sep=",")

Quante righe e colonne sono presenti nei due dataset? Tra le righe sono poi presenti valori nulli?

In [ ]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56707 entries, 0 to 56706
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           56707 non-null  int64  
 1   is_canceled                     56707 non-null  int64  
 2   lead_time                       56707 non-null  int64  
 3   arrival_date_year               56707 non-null  int64  
 4   arrival_date_month              56707 non-null  int64  
 5   arrival_date_week_number        56707 non-null  int64  
 6   arrival_date_day_of_month       56707 non-null  int64  
 7   stays_in_weekend_nights         56707 non-null  int64  
 8   stays_in_week_nights            56707 non-null  int64  
 9   adults                          56707 non-null  int64  
 10  children                        56707 non-null  float64
 11  babies                          56707 non-null  int64  
 12  meal                            

In [ ]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40687 entries, 0 to 40686
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           40687 non-null  int64  
 1   is_canceled                     40687 non-null  int64  
 2   lead_time                       40687 non-null  int64  
 3   arrival_date_year               40687 non-null  int64  
 4   arrival_date_month              40687 non-null  int64  
 5   arrival_date_week_number        40687 non-null  int64  
 6   arrival_date_day_of_month       40687 non-null  int64  
 7   stays_in_weekend_nights         40687 non-null  int64  
 8   stays_in_week_nights            40687 non-null  int64  
 9   adults                          40687 non-null  int64  
 10  children                        40687 non-null  float64
 11  babies                          40687 non-null  int64  
 12  meal                            

In [ ]:
test.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

In [ ]:
train.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
agent                             0
company                           0
days_in_waiting_list              0
customer_type                     0
adr                         

Nel "Train dataset" sono presenti 56707 righe mentre nel "Test Dataset" 40687. Le colonne, invece, sono 30 per entrambi e no in entrambi i dateset non sono presenti valori nulli

I valori sono per lo più numerici o categorici?

In [ ]:
test_non.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40687 entries, 0 to 40686
Data columns (total 30 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   hotel                           40687 non-null  object 
 1   is_canceled                     40687 non-null  int64  
 2   lead_time                       40687 non-null  int64  
 3   arrival_date_year               40687 non-null  int64  
 4   arrival_date_month              40687 non-null  object 
 5   arrival_date_week_number        40687 non-null  int64  
 6   arrival_date_day_of_month       40687 non-null  int64  
 7   stays_in_weekend_nights         40687 non-null  int64  
 8   stays_in_week_nights            40687 non-null  int64  
 9   adults                          40687 non-null  int64  
 10  children                        40687 non-null  float64
 11  babies                          40687 non-null  int64  
 12  meal                            

I valori sono perlopiù categorici, molti dei valori segnati come "int" in realtà sono la conversione in numero di "Si" e "No".

In [ ]:
#numero famiglie con bambini
bambini = train[(train["children"] > 0.00 ) | (train["babies"] > 0.00)]
bambini

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
1          1            0        109               2016                   1   
4          1            1         20               2016                   1   
6          1            0         88               2016                   1   
73         1            1        103               2016                   1   
120        1            1        185               2016                   1   
...      ...          ...        ...                ...                 ...   
56688      0            0        259               2016                  12   
56691      0            0         42               2016                  12   
56692      0            0         23               2016                  12   
56700      0            0        177               2016                  12   
56701      0            0        173               2016                  12   

       arrival_date_week_number  arrival_date_day_of_month  \
1                             1                          1   
4                             1                          1   
6                             1                          1   
73                            2                          9   
120                           5                         28   
...                         ...                        ...   
56688                        53                         28   
56691                        53                         31   
56692                        53                         30   
56700                        53                         30   
56701                        53                         26   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
1                            0                     1       2  ...   
4                            0                     2       2  ...   
6                            0                     2       2  ...   
73                           2                     6       2  ...   
120                          2                     3       2  ...   
...                        ...                   ...     ...  ...   
56688                        2                     5       2  ...   
56691                        2                     2       2  ...   
56692                        2                     3       2  ...   
56700                        2                     5       3  ...   
56701                        3                     8       2  ...   

       assigned_room_type  booking_changes  deposit_type  agent  company  \
1                       7                0             0  240.0     -1.0   
4                       6                0             0  240.0     -1.0   
6                       3                0             0  241.0     -1.0   
73                      6                0             0  240.0     -1.0   
120                     0                0             0  240.0     -1.0   
...                   ...              ...           ...    ...      ...   
56688                   0                3             0    9.0     -1.0   
56691                   0                1             0    9.0     -1.0   
56692                   3                1             0    9.0     -1.0   
56700                   5                0             0   -1.0     -1.0   
56701                   5                4             0   -1.0     -1.0   

       days_in_waiting_list  customer_type     adr  \
1                         0              3  116.10   
4                         0              2  119.00   
6                         0              2   90.17   
73                        0              2   77.40   
120                       0              2   32.40   
...                     ...            ...     ...   
56688                     0              2   80.26   
56691                     0              2  102.00   
56692                     0              2  119.36   
56700                     0    

In [ ]:
#unisco colonne children e babies
train["colonna_bambini"] = train["babies"] + train["children"]
train["colonna_bambini"]

0        0.0
1        2.0
2        0.0
3        0.0
4        2.0
        ... 
56702    0.0
56703    0.0
56704    0.0
56705    0.0
56706    0.0
Name: colonna_bambini, Length: 56707, dtype: float64

In [ ]:
#media cancellazioni in base al numero di bambini
train.groupby(train["colonna_bambini"])["is_canceled"].mean()

colonna_bambini
0.0     0.360020
1.0     0.305078
2.0     0.400452
3.0     0.306122
10.0    0.000000
Name: is_canceled, dtype: float64

In [ ]:
train[train["colonna_bambini"] > 9.00]

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
18765      0            0         37               2016                   1   

       arrival_date_week_number  arrival_date_day_of_month  \
18765                         3                         12   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
18765                        0                     2       2  ...   

       booking_changes  deposit_type  agent  company  days_in_waiting_list  \
18765                1             0    9.0     -1.0                     0   

       customer_type    adr  required_car_parking_spaces  \
18765              2  84.45                            0   

       total_of_special_requests  colonna_bambini  
18765                          1             10.0  

[1 rows x 31 columns]

In [ ]:
train.iloc[18765]

hotel                                0.00
is_canceled                          0.00
lead_time                           37.00
arrival_date_year                 2016.00
arrival_date_month                   1.00
arrival_date_week_number             3.00
arrival_date_day_of_month           12.00
stays_in_weekend_nights              0.00
stays_in_week_nights                 2.00
adults                               2.00
children                             0.00
babies                              10.00
meal                                 0.00
country                            135.00
market_segment                       6.00
distribution_channel                 3.00
is_repeated_guest                    0.00
previous_cancellations               0.00
previous_bookings_not_canceled       0.00
reserved_room_type                   3.00
assigned_room_type                   3.00
booking_changes                      1.00
deposit_type                         0.00
agent                             

In [ ]:
train["persone"] = train["babies"] + train["children"] + train["adults"]
train["persone"]

0        2.0
1        4.0
2        2.0
3        2.0
4        4.0
        ... 
56702    2.0
56703    0.0
56704    2.0
56705    2.0
56706    2.0
Name: persone, Length: 56707, dtype: float64

In [ ]:
train.groupby(train["persone"])["is_canceled"].mean()

persone
0.0     0.154762
1.0     0.301176
2.0     0.380468
3.0     0.305454
4.0     0.405508
5.0     0.254902
12.0    0.000000
Name: is_canceled, dtype: float64

In [ ]:
train[train["persone"] == 0]

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
389        1            1          0               2016                   2   
5638       1            1          0               2016                  11   
18555      1            0         31               2016                  12   
19312      0            1          1               2016                   2   
19326      0            0          0               2016                   2   
...      ...          ...        ...                ...                 ...   
55556      0            0          0               2016                  12   
56000      0            0          0               2016                  12   
56053      0            0          0               2016                  12   
56443      0            0        255               2016                  12   
56703      0            0        177               2016                  12   

       arrival_date_week_number  arrival_date_day_of_month  \
389                           8                         15   
5638                         48                         21   
18555                        53                         27   
19312                         7                         11   
19326                         7                         12   
...                         ...                        ...   
55556                        50                          7   
56000                        52                         18   
56053                        52                         20   
56443                        53                         27   
56703                        53                         30   

       stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
389                          0                     0       0  ...   
5638                         0                     0       0  ...   
18555                        2                     8       0  ...   
19312                        0                     3       0  ...   
19326                        0                     1       0  ...   
...                        ...                   ...     ...  ...   
55556                        0                     0       0  ...   
56000                        0                     0       0  ...   
56053                        0                     0       0  ...   
56443                        0                     3       0  ...   
56703                        2                     5       0  ...   

       deposit_type  agent  company  days_in_waiting_list  customer_type  \
389               0   -1.0    383.0                     0              2   
5638              0   -1.0    386.0                     0              1   
18555             0   -1.0     -1.0                     0              2   
19312             0    9.0     -1.0                     0              2   
19326             0   -1.0     45.0                     0              2   
...             ...    ...      ...                   ...            ...   
55556             0    7.0     -1.0                     0              2   
56000             0    7.0     -1.0                     0              2   
56053             0   28.0     -1.0                     0              2   
56443             0   79.0     -1.0                     0              2   
56703             0   -1.0     -1.0                     0              2   

         adr  required_car_parking_spaces  total_of_special_requests  \
389     0.00                            0                          0   
5638    0.00                            0                          0   
18555  28.00                            1                          0   
19312   0.00                            0                          1   
19326   0.00                            0                          2   
...      ...                          ...                        ...   
55556   0.00                            0                          

Il numero di famiglie (composte da un numero qualsiasi di adulti) con almeno un bambino ("babies" e/o "children") a seguito è di 4496. Il numero massimo di bambini è di 10 raggiunto però in un singolo caso. Escludendo quello il massimo di bambini a seguito è di 3. Le cancellazioni non sembrano a un primo sguardo influenzate dal numero di bambini a seguito. Escludendo suddetto caso estremo il numero massimo di esponenti di una singola famiglia è 5, anche il numero di questi non sembra avere particolare impatto sulle cancellazioni. 84 righe poi presentano il dato anomalo di zero persone (considerare come nullo?).



In [ ]:
#media cancellazioni in base al mese di arrivo
train.groupby(train["arrival_date_month"])["is_canceled"].mean()

arrival_date_month
1     0.247776
2     0.343613
3     0.306177
4     0.379698
5     0.349580
6     0.396070
7     0.327865
8     0.360458
9     0.374861
10    0.405288
11    0.367310
12    0.362176
Name: is_canceled, dtype: float64

Prova 0001

La media delle cancellazioni ha un picco a ottobre mentre raggiunge il valore più basso a gennaio. dividi i mesi per stagioni? (marzo, aprile, maggio = primavera; giugno, luglio, agosto = estate; etc.)